ccmbioinfo/lfs
# CARDS-1737: When computing the surveys that need to be filled out for a visit based on the configured frequency, look at previous visit from the same clinic rather than with the same questionnaire set

[Jira link](https://phenotips.atlassian.net/browse/CARDS-1737).

This PR changes the required forms that need to be filled out for a visit to be linked to the _clinic_, rather than just the questionnaire.

**To test:** Checkout the test branch at `CARDS-1737-test` before continuing, as it contains a change to `mock_torch.js` to allow it to send test data for different clinics.
** Making sure nothing broke: **
1. Under the OSGi config (for me: `http://localhost:8080/system/console/configMgr`), scroll to the bottom and create a new `PROMs import` configuration, with the following:
- Name: test
- Endpoint URL: `http://localhost:8011/`
- Clinic-name: `6012-HC-Congenital Cardiac`
- Everything else default. Save.
2. Startup nodejs mock Torch server from `Utilities/Development` with the command `nodejs mock_torch.js --appointment-time-hours-from-now=48`
3. Open up `http://localhost:8080/Subjects.importTorch.json?config=mock` in your browser. You should see "Data import started". At this point there's an error in the logs because the Authentication URL is broken, but it should otherwise work.
4. Running this import multiple times should still just result in 9 new forms
5. On the sidebar, navigating to PMCC - ACHD and checking Upcoming appointments should result in just 3 of them no matter how many times you ran the import.

![image](https://user-images.githubusercontent.com/4656440/168902953-4a0b47da-2a51-4972-95e9-e736c41a80ee.png)

**Testing the clinic-specific changes:**
6. Go to Administration->Clinics->Create a new Clinic with ID `TestMapping` and name `Test clinic mapping` and Surveys `Mood`. Save.
7. Stop the nodejs mock Torch server, and rerun it specifying your new clinic, a.k.a.: `nodejs mock_torch.js --appointment-time-hours-from-now=48 --appointment-location=TestMapping`
8. Under the OSGi config (for me: `http://localhost:8080/system/console/configMgr`), edit the latest configuration so the Clinic-name is set to `TestMapping`.
9. Open up `http://localhost:8080/Subjects.importTorch.json?config=mock` in your browser. You should see "Data import started".
10. Now there should be 15 forms (Mood has less forms)
11. Under the new Test sidebar entry, 3 new upcoming appointments should be filled out. Make sure the URLs under the Survey completion differ from the ones for PMC - ACHD
![image](https://user-images.githubusercontent.com/4656440/168903854-85894d3b-78d8-4f10-a4ee-7745a2aa8fda.png)

Anything else that looked at the old `surveys` field in `Visit Information` should still be functional, even though the field itself is now gone from imports.
